File: Java/ByTier/E/2127_E_Ancient_Tree.java
// Problem: CF 2127 E - Ancient Tree
// https://codeforces.com/contest/2127/problem/E

/*
 * Problem: E. Ancient Tree
 * Purpose: Given a tree where some vertices have lost their colors (represented by 0),
 * we must assign colors from 1 to k such that the total weight of "cutie" vertices is minimized.
 * A vertex is cutie if there exist two vertices x and y with lca(x,y) = v, 
 * c[x] = c[y], and c[x] != c[v].
 * Algorithm: Use DFS to compute optimal color assignments by propagating color sets from leaves upward,
 * resolving conflicts by choosing colors that minimize cost.
 * Time Complexity: O(n * k) amortized for each test case due to set operations during DFS traversal.
 * Space Complexity: O(n * k) for adjacency list and auxiliary data structures.
 */

import java.util.*;

public class Main {
    static int n, k, a[], color[];
    static long ans;
    static List<List<Integer>> adj;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int test = sc.nextInt();
        while (test-- > 0) {
            n = sc.nextInt();
            k = sc.nextInt();
            a = new int[n];
            color = new int[n];
            adj = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                a[i] = sc.nextInt();
                adj.add(new ArrayList<>());
            }
            int cnt0 = 0;
            for (int i = 0; i < n; i++) {
                color[i] = sc.nextInt();
                if (color[i] == 0) {
                    cnt0++;
                }
            }
            for (int i = 0; i < n - 1; i++) {
                int u = sc.nextInt() - 1;
                int v = sc.nextInt() - 1;
                adj.get(u).add(v);
                adj.get(v).add(u);
            }
            if (cnt0 == n) {
                System.out.println(0);
                for (int i = 1; i <= n; i++) {
                    System.out.print(1 + " ");
                }
                System.out.println();
                continue;
            }
            ans = 0;
            dfs(0, -1); // First DFS to handle color assignments based on subtree information
            dfs2(0, -1, -1); // Second DFS for final consistency of coloring
            System.out.println(ans);
            for (int i = 0; i < n; i++) {
                System.out.print(color[i] + " ");
            }
            System.out.println();
        }
    }

    // DFS to propagate color information and determine optimal assignment
    public static HashSet<Integer> dfs(int u, int p) {
        HashSet<Integer> set = new HashSet<>();
        if (u != 0 && adj.get(u).size() == 1) { // Leaf node
            if (color[u] != 0) {
                set.add(color[u]);
            }
            return set;
        }
        int any = 0;
        HashSet<Integer> matching = new HashSet<>();
        int k = 0;
        for (int v : adj.get(u)) {
            if (v == p) continue;
            HashSet<Integer> set2 = dfs(v, u);
            if (color[v] != 0) {
                any = color[v];
            }
            if (set.size() < set2.size()) {
                HashSet<Integer> temp = set;
                set = set2;
                set2 = temp;
            }
            for (int el : set2) {
                if (set.contains(el)) {
                    matching.add(el);
                    k = el;
                } else {
                    set.add(el);
                }
            }
        }
        // Update node u's color according to conflict resolution strategy
        if (matching.size() == 1) {
            if (color[u] == 0) color[u] = k;
            else if (color[u] != k) ans = ans + a[u];
        } else if (matching.size() == 0) {
            if (color[u] == 0) color[u] = any;
        } else if (matching.size() >= 2) {
            if (color[u] == 0) color[u] = any;
            ans = ans + a[u];
        }
        if (color[u] != 0) set.add(color[u]);
        return set;
    }

    // DFS to assign remaining colors consistently across tree
    public static void dfs2(int u, int p, int c) {
        if (color[u] == 0) { // Propagate color from parent if not yet assigned
            color[u] = c;
        }
        for (int v : adj.get(u)) {
            if (v == p) continue;
            dfs2(v, u, color[u]);
        }
    }
}


// https://github.com/VaHiX/CodeForces/

File: CSharp/ByTier/E/2127_E_Ancient_Tree.cs
// Problem: CF 2127 E - Ancient Tree
// https://codeforces.com/contest/2127/problem/E

﻿/*
E. Ancient Tree
Algorithm: 
  - Uses a tree DFS traversal to compute optimal coloring of vertices to minimize the sum of weights of "cutie" vertices.
  - A vertex v is cutie if there exist x, y such that lca(x, y) = v, c_x = c_y, and c_x != c_v.
  - For each node, we keep track of color sets from its children using a hashset for efficient intersection check.
  - Colors are assigned during DFS traversal to minimize the number of cutie vertices.
Time Complexity: O(n * k) per test case in average case due to hashset operations, where n is number of nodes and k is number of colors.
Space Complexity: O(n * k) for storing graph structure, color assignments, and intermediate hashsets during DFS.

*/

using System.Text;
namespace AncientTree;
class Program {
  static void Solve() {
    int n = Next();
    k = Next();
    w = new long[n];
    for (int i = 0; i < n; i++)
      w[i] = Next();
    c = new int[n];
    for (int i = 0; i < n; i++)
      c[i] = Next();
    nn = new List<int>[n];
    for (int i = 0; i < n; i++)
      nn[i] = new();
    for (int i = 1; i < n; i++) {
      int a = Next() - 1;
      int b = Next() - 1;
      nn[a].Add(b);
      nn[b].Add(a);
    }
    ans = 0;
    h = new HashSet<int>[n];
    Dfs(0, 0); // starts DFS from root (vertex 0)
    if (c[0] == 0) {
      Color(0, 0, 1); // assign color to root if needed
    }
    writer.WriteLine(ans);
    foreach (var i in c) {
      writer.Write(i);
      writer.Write(' ');
    }
    writer.WriteLine();
  }

  static void Color(int to, int from, int color) {
    c[to] = color;
    foreach (var j in nn[to]) {
      if (j == from)
        continue;
      Color(j, to, color); // Recursively color children
    }
  }

  static void Dfs(int to, int from) {
    // Base case - leaf node: recurse down to children
    foreach (var j in nn[to]) {
      if (j == from)
        continue;
      Dfs(j, to);
    }
    
    // Find the child with maximum number of distinct colors for optimization
    int max = -1;
    foreach (var j in nn[to]) {
      if (j == from)
        continue;
      if (max == -1 || h[max].Count < h[j].Count)
        max = j;
    }
    
    // Initialize the set of colors for current node
    if (max == -1) {
      h[to] = new();
      if (c[to] != 0)
        h[to].Add(c[to]);
      return;
    }
    
    // Copy the color set from the max child to current
    h[to] = h[max];
    
    // If current node has no assigned color, proceed with logic for determining its color
    if (h[to].Count == 0) {
      if (c[to] != 0) {
        h[to].Add(c[to]);
        Color(to, from, c[to]); // Color subtree of this node
      }
      return;
    }
    
    HashSet<int> cutie = new(); // track colors that would make parent a cutie
    foreach (var j in nn[to]) {
      if (j == from || j == max)
        continue;
      foreach (var y in h[j]) {
        if (!h[to].Add(y)) // Add color only if it's not already present, indicating a conflict
          cutie.Add(y);
      }
    }
    
    // Determine cost and coloring strategy based on number of conflicting colors
    if (cutie.Count > 1) {
      ans += w[to];
      if (c[to] == 0)
        c[to] = cutie.First(); // assign one conflicting color
    } else if (cutie.Count == 1) {
      if (c[to] == 0) {
        c[to] = cutie.First(); // assign the single conflicting color
      } else if (c[to] != cutie.First()) {
        ans += w[to]; // add cost because color conflicts with what was chosen for parent
      }
    } else {
      if (c[to] == 0)
        c[to] = h[to].First(); // assign any available color from subtree
    }

    h[to].Add(c[to]); // add current node's color to color set
    
    // Assign color to non-colored descendants using same chosen color as current node
    foreach (var j in nn[to]) {
      if (j == from)
        continue;
      if (c[j] == 0)
        Color(j, to, c[to]);
    }
  }

  static int k;
  static HashSet<int>[] h;
  static long ans;
  static long[] w;
  static int[] c;
  static List<int>[] nn;

  static void Main(string[] args) {
    long t = Next();
    for (int i = 0; i < t; i++)
      Solve();
    writer.Flush();
  }

  static int Next() {
    int c;
    int m = 1;
    int res = 0;
    do {
      c = reader.Read();
      if (c == '-')
        m = -1;
    } while (c < '0' || c > '9');
    res = c - '0';
    while (true) {
      c = reader.Read();
      if (c < '0' || c > '9')
        return m * res;
      res *= 10;
      res += c - '0';
    }
  }

  static readonly StreamReader reader = new(
      Console.OpenStandardInput(1024 * 10), Encoding.ASCII, false, 1024 * 10);
  static readonly StreamWriter writer =
      new(Console.OpenStandardOutput(1024 * 10), Encoding.ASCII, 1024 * 10);
}


// https://github.com/VaHiX/codeForces/